st3/sublime_lib/resource_path.py

```diff
@@ -97,46 +224,43 @@ def from_file_path(cls, file_path):
            )
            ResourcePath("Packages/My Package/foo.py")
         """
-        file_path = Path(file_path)
-        if not file_path.is_absolute():
-            raise ValueError("Cannot convert a relative file path to a resource path.")
 
-        for root, base in get_resource_roots().items():
-            try:
-                rel = file_path.relative_to(base)
-            except ValueError:
-                pass
-            else:
-                return cls(root, *rel.parts)
-
-        for base in get_installed_resource_roots():
-            try:
-                rel = file_path.relative_to(base).parts
-
-                if rel == ():
-                    return cls('Packages')
-                else:
-                    package, *rest = rel
-                    return (cls('Packages', package)
-                            .remove_suffix('.sublime-package').joinpath(*rest))
-            except ValueError:
-                pass
-
-        raise ValueError("Path {!r} does not correspond to any resource path.".format(file_path))
+        file_path = wrap_path(file_path)
+        try:
+            return next(
+                path
+                for path in (
+                    root.file_to_resource_path(file_path)
+                    for root in get_roots()
+                )
+                if path is not None
+            )
+        except StopIteration:
+            raise ValueError(
+                "Path {!r} does not correspond to any resource path.".format(file_path)
+            )
```

Comment:
```suggestion
            ) from None
```

I don't like this section. One one hand, it's ingenious (seemingly), but on the other hand it is error-prone (like I showed) and it's hard to get the order of execution right in your head. The double comprehension is boilerplatey.

```py
candidates = root.file_to_resource_path(file_path) for root in get_roots()
path = next(filter(None, candidates), None)
if path:
    return path
else:
    raise ValueError(
        "Path {!r} does not correspond to any resource path.".format(file_path)
    )
```
is easier to read, has less lines and is more explicit.

Reply:
Done.